doc: update quick start for verifying using ts/tp (#82)
notaryproject/notaryproject.dev: content/en/docs/quickstart.md

> ```console

Reviewer: @yizha1  I think we should also claim that this quick start is using Notation v0.12.0-beta.1 at the beginning since the Notary doc has not introduced version control yet. I see Zach was confused about the testing result when using an old version of Notation.

Author: > @yizha1 I think we should also claim that this quick start is using Notation v0.12.0-beta.1 at the beginning since the Notary doc has not introduced version control yet. I see Zach was confused about the testing result when using an old version of Notation. 

The installation is moved to another page if you check the "Prerequisites". Maybe this was done by Zach.  I think there is a reason behind this, which is always to use the latest version. In hello-signing.md, there is version info.

```

## Sign the container image

Use `notation sign` to sign the container image with COSE signature format.

```console
notation sign --plain-http --signature-format cose $IMAGE
```

Use `notation list` to show the signatures associated with the container image.

```console
notation list --plain-http $IMAGE
```

Confirm there is one signature, for example:

```output
$ notation list --plain-http $IMAGE
localhost:5000/net-monitor@sha256:073b75987e95b89f187a89809f08a32033972bb63cda279db8a9ca16b7ff555a
└── application/vnd.cncf.notary.v2.signature
    └── sha256:ba3a68a28648ba18c51a479145fca60d96b43dc96c6ab22f412c89ac56a9038b
```

## Create a trust policy

In order to verify the container image, you need to configure the trust policy to specify trusted identities which sign the artifacts, and level of signature verification to use. See [trust policy spec](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-policy) to understand more about trust policy.

The trust policy is a JSON document with the name `trustpolicy.json` stored in notation configuration directory.

For a Linux user, the directory is `$HOME/.config/notation/`.

> For a Window user, the directory is `C:/Users/AppData/Roaming/notation/`.

Reviewer 2: Missing username after user

For linux user,

```console
cat <<EOF > "$HOME/.config/notation/trustpolicy.json"
{
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "wabbit-networks-images",
            "registryScopes": [ "*" ],
            "signatureVerification": {
                "level" : "strict" 
            },
            "trustStores": [ "ca:wabbit-networks.io" ],
            "trustedIdentities": [
                "*"
            ]
        }
    ]
}
EOF
```

For Mac user,

```console
cat <<EOF > "$HOME/Library/Application Support/notation/trustpolicy.json"
{
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "wabbit-networks-images",
            "registryScopes": [ "*" ],
            "signatureVerification": {
                "level" : "strict" 
            },
            "trustStores": [ "ca:wabbit-networks.io" ],
            "trustedIdentities": [
                "*"
            ]
        }
    ]
}
EOF
```

For Windows user, create a json file named `trustpolicy.json` with the following content, and store it under `C:/Users/AppData/Roaming/notation/`

> 

Reviewer 2: I'd create one policy file in the doc and have different os version home directories in one spot

Confirm there are no signatures.

## Generate a test key and self-signed certificate

Use `notation cert generate-test` to generate a test RSA key for signing artifacts, and a self-signed test certificate for verifying artifacts.

**IMPORTANT**: Self-signed certificates should be used for development purposes only and should not be used in production environments.

The following command generates a test key and a self-signed X.509 certificate under the `$HOME/.config/notation/localkeys` directory. The test key is set as a default signing key using `--default` flag. The self-signed certificate is added to a named trust store `wabbit-networks.io` of type `ca`.

```console
notation cert generate-test --default "wabbit-networks.io"
```

Use `notation key list` to confirm the signing key. Key name with a `*` prefix is the default key.

```console
notation key list
```

> Use `notation certificate list` to confirm the certificate is stored in the trust store.

Reviewer: Use `notation certificate list` to confirm whether the certificate is stored in the trust store.

Reviewer 3: also, on smoke test, the command shows no certificates added to the store

Author: > also, on smoke test, the command shows no certificates added to the store

Please use the new release https://github.com/notaryproject/notation/releases/tag/v0.12.0-beta.1 @zr-msft

```console
notation cert list
```

## Sign the container image

Use `notation sign` to sign the container image with COSE signature format.

```console

> notation sign --plain-http --signature-format cose $IMAGE

Reviewer 3: this command fails for me 

```
notation sign --plain-http --signature-format cose $IMAGE
Error: unknown flag: --signature-format
2022/11/01 15:19:08 unknown flag: --signature-format
```

Author: > this command fails for me
> 
> ```
> notation sign --plain-http --signature-format cose $IMAGE
> Error: unknown flag: --signature-format
> 2022/11/01 15:19:08 unknown flag: --signature-format
> ```

Sorry. This is for an upcoming notation release beta.1. The version is not ready yet. It should be ready by Nov 2 PDT.

```console
notation cert generate-test --default "wabbit-networks.io"
```

Use `notation key list` to confirm the signing key. Key name with a `*` prefix is the default key.

```console
notation key list
```

Use `notation certificate list` to confirm whether the certificate is stored in the trust store.

```console
notation cert list
```

## Sign the container image

Use `notation sign` to sign the container image with COSE signature format.

```console
notation sign --plain-http --signature-format cose $IMAGE
```

Use `notation list` to show the signatures associated with the container image.

```console
notation list --plain-http $IMAGE
```

Confirm there is one signature, for example:

```output
$ notation list --plain-http $IMAGE
localhost:5000/net-monitor@sha256:073b75987e95b89f187a89809f08a32033972bb63cda279db8a9ca16b7ff555a
└── application/vnd.cncf.notary.v2.signature
    └── sha256:ba3a68a28648ba18c51a479145fca60d96b43dc96c6ab22f412c89ac56a9038b
```

## Create a trust policy

> In order to verify the container image, you need to configure the trust policy to specify trusted identities which sign the artifacts, and level of signature verification to use. The trust policy is a JSON document with the name `trustpolicy.json` under `$HOME/.config/notation/`.

Reviewer 2: Provide link to local filesystem document since $HOME doesn't work for everyone.  Also, might be a "cheat" to grab that path from the generated certificate with `notation cert ls`

> Once the trust policy file is executed as below, it creates one trust policy named `wabbit-networks-images`. This policy only applies to the repository `$REPO`. The level of signatureVerification is `strict`, which enforces a full validation. Any validation failure will fail the whole signature verification process. The trust store used for this policy is named `wabbit-networks.io` of type `ca`, which is created in previous step. Use `notation cert show` to get trust identity info from the subject field of the self-signed certificate.

Reviewer 2: Add link to more details on trust policy semantics doc.

```console

> cat <<EOF > $HOME/.config/notation/trustpolicy.json

Reviewer 4: This doesn't seem to work on a Mac?

```
cat <<EOF > $HOME/.config/notation/trustpolicy.json
{
<omitted>
}
EOC
```

```
notation verify <image>:1.25.4
Error: open /Users/jrrickard/Library/Application Support/notation/trustpolicy.json: no such file or directory
```

{
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "wabbit-networks-images",

>             "registryScopes": [ "$REPO" ],

Reviewer 4: This should be explicit. Is it the registry? Is it the repo minus an image? Does someone have to specify this for every image they want to validate?

Reviewer 4: For what it's worth, with the current version I had to include the image specifically:

```
cat <<EOF > $HOME/.config/notation/trustpolicy.json
{
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "upstream",
            "registryScopes": [ "upstream.azurecr.io/oss/fluxcd/flux" ],
            "signatureVerification": {
                "level" : "strict" 
            },
            "trustStores": [ "ca:trust" ],
            "trustedIdentities": [
                "*"
            ]
        }
    ]
}
EOF
```

If I omitted "flux" I ended up getting errors:

```
ERROR: artifact "upstream.azurecr.io/oss/fluxcd/flux@sha256:0f9ad5df470963f959bb3e0b7abe60e9b4011a636e566a06e883ed7b3812dbc4" has no applicable trust policy

Signature verification failed for all the 0 signatures associated with digest: sha256:0f9ad5df470963f959bb3e0b7abe60e9b4011a636e566a06e883ed7b3812dbc4

Error: artifact "upstream.azurecr.io/oss/fluxcd/flux@sha256:0f9ad5df470963f959bb3e0b7abe60e9b4011a636e566a06e883ed7b3812dbc4" has no applicable trust policy
```

You should allow wild cards here or something. From an end user perspective if I have to build a trust policy that includes every image I want to validate, that becomes an issue when I build any additional images and is a cumbersome experience.

Reviewer 2: `upstream.azurecr.io/*` should work and `["$REPOSITORY/*"]` should be in the example

Reviewer 4: I’ll give that a try!

Author: According to the [spec](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#registry-scopes-constraints), the repository URI MUST NOT contain the asterisk character *. You can just use `["*"]` as a global scope for a quick start.

Reviewer 4: Thanks for the additional clarification. So in summary:

Option 1: Trust everything 
Option 2: enumerate every image you will verify?

Author: > Thanks for the additional clarification. So in summary:
> 
> Option 1: Trust everything Option 2: enumerate every image you will verify?

@jeremyrickard , you can specify a repo, then all the artifacts under the repo will be verified against the policy. For example, you have images under repo `localhost:5000/net-monitor`
- localhost:5000/net-monitor:v1
- localhost:5000/net-monitor:v2
- localhost:5000/net-monitor:v3

you just need to specify `localhost:5000/net-monitor` for registryScopes parameter.

Author: Check out here for more scenarios: https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-policy-schema

Reviewer 2: @yizha1 - yes this is far too much work!  We need to add a backlog item to support all images coming from a registry.

            "signatureVerification": {
                "level" : "strict" 
            },
            "trustStores": [ "ca:wabbit-networks.io" ],
            "trustedIdentities": [

>                 "x509.subject: CN=wabbit-networks.io,O=Notary,L=Seattle,ST=WA,C=US"

Reviewer 2: Please use: `"trustedIdentities": ["*"]` instead so it is an easier reference for people to build.

Confirm there are no signatures.

## Generate a test key and self-signed certificate

Use `notation cert generate-test` to generate a test RSA key for signing artifacts, and a self-signed test certificate for verifying artifacts.

**IMPORTANT**: Self-signed certificates should be used for development purposes only and should not be used in production environments.

> The following command generates a test key and a self-signed X.509 certificate under the `$HOME/.config/notation/localkeys` directory. The test key is set as a default signing key using `--default` flag. The self-signed certificate is added to a named trust store `wabbit-networks.io` of type `ca`.

Reviewer: If we use `$HOME/.config/notation/localkeys` here, I suggest adding a prerequisite to claiming that this Quick Start is using a Linux environment as an example. Since this path is different per OS.

Reviewer: We can also add an inner link to [another document](https://github.com/notaryproject/notaryproject.dev/pull/80/files) that introduces the path when it is merged into the website.

```console
notation cert generate-test --default "wabbit-networks.io"
```

Use `notation key list` to confirm the signing key. Key name with a `*` prefix is the default key.

```console
notation key list
```

> Use `notation certificate list` to confirm whether the certificate is stored in the trust store.

Reviewer: Suggest `notation certificate list` => `notation cert list` to make it consistent with the following script example.

```console
notation cert list
```

## Sign the container image

Use `notation sign` to sign the container image with COSE signature format.

```console

> notation sign --plain-http --signature-format cose $IMAGE

Reviewer: I think `--plain-http` is not a MUST in this case. Can we remove it to keep the signing simpler?

Reviewer 5: We can remove remove `--signature-format cose`

```

Use `notation list` to show the signatures associated with the container image.

```console
notation list --plain-http $IMAGE
```

Confirm there is one signature, for example:

```output
$ notation list --plain-http $IMAGE
localhost:5000/net-monitor@sha256:073b75987e95b89f187a89809f08a32033972bb63cda279db8a9ca16b7ff555a
└── application/vnd.cncf.notary.v2.signature
    └── sha256:ba3a68a28648ba18c51a479145fca60d96b43dc96c6ab22f412c89ac56a9038b
```

## Create a trust policy

> In order to verify the container image, you need to configure the trust policy to specify trusted identities which sign the artifacts, and level of signature verification to use. See [trust policy spec](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-policy) to understand more about trust policy.

Reviewer: There is a [Trust Policy doc](https://notaryproject.dev/docs/concepts/trust-store-trust-policy-specification/#trust-policy) on the Notary website already, we should use a relative link here. You can replace it with `[trust policy spec]({{< ref "/docs/concepts/trust-store-trust-policy-specification/#trust-policy" >}})`.

> Create a json file named `trustpolicy.json` with the following content:

Reviewer: json => JSON

```json
{
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "wabbit-networks-images",
            "registryScopes": [ "*" ],
            "signatureVerification": {
                "level" : "strict" 
            },
            "trustStores": [ "ca:wabbit-networks.io" ],
            "trustedIdentities": [
                "*"
            ]
        }
    ]
}
```

For a Linux user, store file `trustpolicy.json` under directory `$HOME/.config/notation/`.

For a Mac user, store file `trustpolicy.json` under directory `$HOME/Library/Application Support/notation/`.

For a Window user, store file `trustpolicy.json` under directory `C:\Users\<username>\AppData\Roaming\notation\`.

> Once the trust policy file is executed, it creates one trust policy named `wabbit-networks-images`. The trust policy with `registryScopes` set to `*` applies to all the artifacts in any registries. The level of signatureVerification is `strict`, which enforces a full validation. Any validation failure will fail the whole signature verification process. The trust store used for this policy is named `wabbit-networks.io` of type `ca`, which is created in previous step. See [trust policy spec](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-policy-properties) to fine tune the policies for specific security requirements.

Reviewer: Once the trust policy file is created?

> Once the trust policy file is executed, it creates one trust policy named `wabbit-networks-images`. The trust policy with `registryScopes` set to `*` applies to all the artifacts in any registries. The level of signatureVerification is `strict`, which enforces a full validation. Any validation failure will fail the whole signature verification process. The trust store used for this policy is named `wabbit-networks.io` of type `ca`, which is created in previous step. See [trust policy spec](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-policy-properties) to fine tune the policies for specific security requirements.

Reviewer: Same as above. We should use a relative link here. You can replace it with [trust policy spec]({{< ref "/docs/concepts/trust-store-trust-policy-specification/#trust-policy-properties" >}}).

> Once the trust policy file is executed, it creates one trust policy named `wabbit-networks-images`. The trust policy with `registryScopes` set to `*` applies to all the artifacts in any registries. The level of signatureVerification is `strict`, which enforces a full validation. Any validation failure will fail the whole signature verification process. The trust store used for this policy is named `wabbit-networks.io` of type `ca`, which is created in previous step. See [trust policy spec](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-policy-properties) to fine tune the policies for specific security requirements.

Reviewer 5: ```suggestion
The above trustpolicy with the name 'wabbit-networks-images' has `registryScopes` set to `*` which means it applies to all the artifacts of any registry. The `signatureVerification` is set to `strict` which means notation will perform all the validations and any failure in validation will lead to the failure of signature verification. This policy uses the` wabbit-networks.io` truststore of type `ca` which was created in the previous step. For more details please read [trust policy specification](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-policy-properties).
```

## Verify the container image

> Use `notification verify` to verify any signatures on your container image.

Reviewer 5: ```suggestion
Use `notification verify` to verify signatures associated with the container image.
```

```console
notation verify --plain-http $IMAGE
```

## Next steps

For additional examples on using Notary, see:

> * [Build, sign, and verify container images using Notary and Azure Key Vault](https://learn.microsoft.com/azure/container-registry/container-registry-tutorial-sign-build-push)

Reviewer: Is this Ratify guide working with Notation 0.12.0-beta.1?

Author: > Is this Ratify guide working with Notation 0.12.0-beta.1?

Maybe we should delete this link. This is not added by this PR.
